src/ophyd_async/core/mock_signal_utils.py

```diff
@@ -0,0 +1,183 @@
+import asyncio
+import itertools
+from contextlib import asynccontextmanager, contextmanager
+from typing import Any, Callable, Dict, Generator, Iterator, List
+from unittest.mock import ANY
+
+from ophyd_async.core.signal import Signal
+from ophyd_async.core.utils import ReadingValueCallback, T
+
+from .mock_signal_backend import MockSignalBackend
+
+
+def _get_mock_signal_backend(signal: Signal) -> MockSignalBackend:
+    assert isinstance(signal._backend, MockSignalBackend), (
+        "Expected to receive a `MockSignalBackend`, instead "
+        f" received {type(signal._backend)}. "
+    )
+    return signal._backend
+
+
+def set_mock_value(signal: Signal[T], value: T):
+    """Set the value of a signal that is in mock mode."""
+    backend = _get_mock_signal_backend(signal)
+    backend._set_value(value)
+
+
+def set_mock_put_proceeds(signal: Signal[T], proceeds: bool):
+    """Allow or block a put with wait=True from proceeding"""
+    backend = _get_mock_signal_backend(signal)
+
+    if proceeds:
+        backend.put_proceeds.set()
+    else:
+        backend.put_proceeds.clear()
+
+
+def set_mock_callback(signal: Signal[T], callback: ReadingValueCallback[T]) -> None:
+    """Monitor the value of a signal that is in mock mode"""
+
+    backend = _get_mock_signal_backend(signal)
+    return backend.set_callback(callback)
+
+
+# Gives the signal put wait_for a chance to complete
+_A_BIT = 1e-4
+
+
+@asynccontextmanager
+async def block_puts_during(*signals: List[Signal]):
+    for signal in signals:
+        set_mock_put_proceeds(signal, False)
+    yield
+    for signal in signals:
+        set_mock_put_proceeds(signal, True)
+
+    await asyncio.sleep(_A_BIT)
+
+
+@asynccontextmanager
+async def set_signals_after(signals_to_value: Dict[Signal, Any]):
+    async with block_puts_during(*signals_to_value.keys()):
+        yield
+    for signal, value in signals_to_value.items():
+        await signal.set(value)
+
+
+def assert_mock_put_called_with(signal: Signal, value: Any, wait=ANY, timeout=ANY):
+    backend = _get_mock_signal_backend(signal)
+    backend.mock.put.assert_called_with(value, wait=wait, timeout=timeout)
+
+
+class _SetValuesIterator:
+    # Garbage collected by the time __del__ is called unless we put it as a
+    # global attrbute here.
+    require_all_consumed: bool = False
+
+    def __init__(
+        self,
+        signal: Signal,
+        *values: List[Any],
+        repeat: bool = False,
```

Comment:
I can't actually think of a usecase for `repeat` now, as we will be consuming from within a for loop we will always repeat the same value. Please can you delete it?